### PR TITLE
Remove ordereddict io.ascii

### PR DESCRIPTION
**Pull request by MridulS**
### Description

This builds on top of https://github.com/astropy/astropy/pull/15496, the commit https://github.com/astropy/astropy/commit/085fcd0da3faa88e5a8764fd9238f304e4f199ef has the changes to make the tests happy. One is changing the `MetaData` default factory to a `dict` instead of `OrderedDict` and other (more controversial) one is to change the YAML representation of the dump. From what I gather the main reason for the failure was the addition of the `omap` tags which are added by custom astropy dumpers/loaders for `OrderedDict` and not for a normal python `dict`.

But this is a breaking change as it does change the YAML representation, so if someone reads an old serialized object it will be read in a normal dictionary and not a ordered dict. If we do tweak the representation we should get rid of the custom logic for handling ordereddicts in `AstropyDumper/Loader`.

This also forces all the dictionary objects to preserve insertion order in the YAML output, which wasn't guaranteed before.

I think it may make sense to do `astropy/table` in this PR too to make sure everything is in one place.

Fixes #<Issue Number>

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by neutrinoceros** [comment]
If it's building on top of a draft PR I suggest drafting this one too !

**Comment by taldcroft** [comment]
@MridulS - the change to `MetaData` is likely to cause too many problems. The YAML representation of this is bigger than just astropy. A few things to note:
- In YAML the key order is explicitly not ordered. So even if one Python implementation happens to write them out in order this does not guarantee anything with other readers.  https://yaml.org/spec/1.2.2/#3221-mapping-key-order
- The ECSV spec explicitly calls out `!!omap` as the serialization type for `meta`. This cannot be changed without a change to APE-6. There are other non-Python reader/writers for ECSV so such a change is not likely to be approved unless there is a very strong driver. https://github.com/astropy/astropy-APEs/blob/main/APE6.rst

**Comment by MridulS** [comment]
> The ECSV spec explicitly calls out !!omap as the serialization type for meta. This cannot be changed without a change to APE-6. There are other non-Python reader/writers for ECSV so such a change is not likely to be approved unless there is a very strong driver. https://github.com/astropy/astropy-APEs/blob/main/APE6.rst


Woops, did not see that! I guess OrderedDict needs to be there. I don't think it is going away from CPython anytime soon. The only main gain here would be the performance gain. Vanilla dict is supposed to be much more efficient than OrderedDict

**Comment by taldcroft** [comment]
> The only main gain here would be the performance gain. Vanilla dict is supposed to be much more efficient than OrderedDict

I haven't done profiling but I would be a bit surprised if `meta` handling is a substantial factor in performance for many real-world applications.

**Comment by MridulS** [comment]
> I haven't done profiling but I would be a bit surprised if meta handling is a substantial factor in performance for many real-world applications.

Yeah I was thinking more about Table wide usage of OrderedDict, like in `TableColumns` too, but maybe the abstraction is far enough. I should run some benchmarks :)

File: astropy/cosmology/_io/ecsv.py
Comment (nstarman) [inline astropy/cosmology/_io/ecsv.py:1]: I'm fine with all the cosmology changes so long as the resulting files adhere to the ECSV standard!

**Comment by taldcroft** [comment]
@MridulS - I was poking around the code and wondered about an idea to use `OrderedDict` *only* at the interface with YAML, in that way preserving the `!!omap` tag. I have not tested this, but you could try these changes:
```
➜  astropy git:(main) ✗ git diff
diff --git a/astropy/table/meta.py b/astropy/table/meta.py
index 3af1c64cab..20883641c5 100644
--- a/astropy/table/meta.py
+++ b/astropy/table/meta.py
@@ -76,7 +76,7 @@ def _construct_odict(load, node):
       >>> yaml.load('''!!omap [ foo: bar, mumble: quux, baz : gorp ]''')  # doctest: +SKIP
       OrderedDict([('foo', 'bar'), ('mumble', 'quux'), ('baz', 'gorp')])
     """
-    omap = OrderedDict()
+    omap = {}
     yield omap
     if not isinstance(node, yaml.SequenceNode):
         raise yaml.constructor.ConstructorError(
@@ -269,7 +269,7 @@ def _get_col_attributes(col):
         ("unit", lambda x: x is not None, str),
         ("format", lambda x: x is not None, None),
         ("description", lambda x: x is not None, None),
-        ("meta", lambda x: x, None),
+        ("meta", lambda x: OrderedDict(x), None),
     ):
         col_attr = getattr(col.info, attr)
         if nontrivial(col_attr):
@@ -302,7 +302,7 @@ def get_yaml_from_table(table):
     """
     header = {"cols": list(table.columns.values())}
     if table.meta:
-        header["meta"] = table.meta
+        header["meta"] = OrderedDict(table.meta)
 
     return get_yaml_from_header(header)
```

**Comment by taldcroft** [comment]
With the changes above, you should also revert the changes that add `sort_keys=False`. That should no longer be necessary and this will hopefully reduce the change footprint even further. The goal is for this PR to be minimally impacting and not break any unit or doc tests. (That's the goal, we'll see about reality).

File: astropy/cosmology/_io/ecsv.py [hunk 1]
```diff
@@ -24,8 +24,8 @@
     # datatype:
     # - {name: name, datatype: string}
     ...
-    # meta: !!omap
```
Comment (pllim) [inline astropy/cosmology/_io/ecsv.py:27]: What does this mean for people with older ESCV files with this `!!omap` though? Will their files no longer load with this patch? I think this was discussed before elsewhere but I don't have time to dig it up right now.

Comment (MridulS) [reply astropy/cosmology/_io/ecsv.py:27]: Yeah removing that was a bad idea 😅 

This is continued in https://github.com/astropy/astropy/pull/16250 (closing this PR).

**Pull request closed without merging**